orttraining/orttraining/core/framework/checkpointing.cc

```diff
@@ -125,48 +52,77 @@ std::vector<std::string> GetOrderedOrtValueNames(const NameMLValMap& name_to_val
   return ordered_names;
 }
 
+}  // namespace
+
 Status SaveRuntimeTensors(
     const PathString& tensors_path,
     const PathString& tensors_data_path,
     const DataTransferManager& data_transfer_manager,
-    const NameMLValMap& ort_values) {
+    const NameMLValMap& ort_values,
+    bool force_save_as_external_data) {
   // just write data file basename to TensorProto - this will get overwritten
   //   with the actual path when loading the checkpoint
   const PathString tensors_data_relative_path = GetLastComponent(tensors_data_path);
-
   const std::vector<std::string> ordered_tensor_names = GetOrderedOrtValueNames(ort_values);
+  bool save_as_external_data = force_save_as_external_data;
+
+  // Check whether there is potential protobuf size limit was hit.
+  if (!save_as_external_data) {
+    unsigned long total_bytes = 0;
+    constexpr unsigned long PROTOBUF_UPPER_LIMIT = 2 * 1000 * 1000 * 1000;
+    for (const auto& tensor_name : ordered_tensor_names) {
+      const OrtValue& ort_value = ort_values.at(tensor_name);
+      ORT_RETURN_IF_NOT(ort_value.IsTensor(), "ort_value.IsTensor() was false");
+      const Tensor& src_tensor = ort_value.Get<Tensor>();
+      total_bytes += src_tensor.SizeInBytes();
+
+      if (total_bytes >= PROTOBUF_UPPER_LIMIT) {
+        save_as_external_data = true;
+        break;
+      }
+    }
+  }
+
+  // Convert Tensor to TensorProto.
   std::vector<char> tensor_data_buffer{};
   static const OrtMemoryInfo cpu_alloc_info{onnxruntime::CPU, OrtDeviceAllocator};
   std::vector<ONNX_NAMESPACE::TensorProto> saved_tensor_protos{};
   saved_tensor_protos.reserve(ordered_tensor_names.size());
-  std::ofstream tensors_data_file{tensors_data_path};
 
   for (const auto& tensor_name : ordered_tensor_names) {
     const OrtValue& ort_value = ort_values.at(tensor_name);
     ORT_RETURN_IF_NOT(ort_value.IsTensor(), "ort_value.IsTensor() was false");
-    const Tensor& tensor = ort_value.Get<Tensor>();
-
-    tensor_data_buffer.resize(tensor.SizeInBytes());
-    ORT_RETURN_IF_ERROR(CopyTensorDataToByteSpan(
-        data_transfer_manager, tensor, cpu_alloc_info, gsl::make_span(tensor_data_buffer)));
-
-    saved_tensor_protos.emplace_back();
-    ORT_RETURN_IF_ERROR(SaveRuntimeTensor(
-        tensor_name, tensor, tensor_data_buffer, tensors_data_relative_path,
-        tensors_data_file, saved_tensor_protos.back()));
+    const Tensor& src_tensor = ort_value.Get<Tensor>();
+    tensor_data_buffer.resize(src_tensor.SizeInBytes());
+    auto& tensor_location = src_tensor.Location();
+
+    if (tensor_location.device.Type() == OrtDevice::CPU ||
```

Comment:
why are we copying even if the device is CPU?

Reply:
I think we should avoid operating the CPU Tensor in the session state. We can revisit here to optimizer if we find this bring much overhead.